### DISCUSSION TITLE
Why reasonable rules can create infinite loops

### DISCUSSION THREAD

**Discussion by jkoppel** [egraphs-good/egg#60] [E-graph theory]
# Why reasonable rules can create infinite loops

Myth: If a set of rewrite rules terminates when applied to all inputs when viewed as a term-rewriting system, then it will also terminate when used in equality saturation  / e-graph rewriting.

Fact: A set of terminating rewriting rules may yield an infinite number of distinct e-classes.

-----------------------

This post summarizes discussions between myself and several of the `egg` authors about unexpected loops in equality saturation, along with my own attempts at characterizing the problem using term-rewriting theory.

----------------------

## Preamble: Terminology and Background

We introduce some terminology from term-rewriting theory. All of this can be found in either "Term Rewriting and All That"  (Baader and Nipkow, 1998) or "Termination of Rewriting" (Dershowitz, 1987).

### Terminology

**Terminating**: A set of rules is terminating if there is no infinite derivation from any start term.
**Quasi-terminating**: A set of rules is quasi-terminating if every infinite derivation has a cycle
**Globally finite**: A set of rules is globally finite if the number of distinct terms generated from any start term is finite. For systems with a finite number of rules, this is equivalent to quasi-termination.

Note that the property of interest for equality saturation is distinct from global finiteness: equality saturation terminates not if there are a finite number of distinct (non-equivalent) terms, but if there are a finite number of distinct _subterms_. For instance, in the infinite sequence of terms {`0`, `a-a`, `(a+a)+(-a-a)`, `(a+a+a)+(-a-a-a)`, ...}, every term is equivalent to 0, but there are an infinite number of non-equivalent subterms.

**Overlap**: Two patterns `A` and `B` overlap if `B` unifies with a (non-variable) subterm of `A`, or vice versa. See examples below under **critical overlap**.

**Critical overlap**: Two rules `A => B` and `C => D` have _critical overlap_ if `A` unifies with a nontrivial subterm of `B` or vice-versa. For example, the rules `f(g(x)) => A` and `g(h(y)) => B` have critical overlap because `g(x)` unifies with `g(h(y))` (with the unifier `[x |-> h(y)]`). This yields the term `f(g(h(y))`, which steps to both `A` and `f(B)`. The rules `f(x) => A` and `g(y) => B` do not have critical overlap; while `g(y)` does unify with `x` (unifier: `[x |-> g(y)]`), this unification is trivial. The associativity rule `(x+y)+z => x+(y+z)` has critical overlap with itself (unifier: `[x |-> x + y, y |-> z]`), yielding the term after renaming  `((w+x)+y)+z`, which steps to both `(w+x)+(y+z)` and `(w+(x+y))+z`.

### Proving Termination

https://en.wikipedia.org/wiki/Rewrite_order

The typical way to prove termination is to find a _reduction order_ `<=`.  If `<=` is a well-founded partial order and, for any terms `t`, `u` with `t => u`, it must be that `u < t`, then the rewrite system is terminating.

To show quasi-termination, a thin preorder is used instead of a partial order. A preorder is any transitive relation `R`; if `a R b` and `b R a`, then `a` and `b` are in the same equivalence class of `R`. A thin preorder is a preorder whose equivalence classes are all finite.

## Nonterminating Example 

The most common problematic example comes from associativity combined with annihilation.

Consider the system with rules `(a+b)+c => a+(b+c)`, `a+0 => a`, and `(-b)+b => 0`. It is easy to see this ruleset is terminating as a term-rewriting system because each reduction either reduces the lexicographic ordering (total term size, term size of left child). However, when applied to egraphs, it deduces that `a=(a+(-b))+b`, and then produces the infinite derivation

    (a+(-b))+b
    => a+(-b+b)
    = ((a+(-b))+b)+(-b+b)
    => (a+(-b))+(b+(-b+b))
    => a + ((-b)+(b+(-b+b)))
    => ...

For this example, judiciously collapsing all right subterms into the two eclasses of `b` and `0` prevents blowup. However, merely adding the commutativity rule `a+b => b+a` (which brings the underlying rewrite system from terminating to quasi-terminating) permits the system to rearrange the arbitrarily-large sum of `b`'s and `-b`'s arbitrarily, resulting in an infinite number of eclasses, and hence causes equality saturation to loop infinitely.

## Characterization

Precisely characterizing the situations when e-graph rewriting blows up for a "reasonable" rule-set (or, more generally, characterizing the differences between e-graph and term rewriting)  is an unsolved research problem. However, here are some properties of an offending rule system:

### Cyclic terms are bad

In the associativity example, the problem came that `(a+(-b)+b` is equivalent to a subterm of itself, `a`. Replacing `a` with the identifier "eclass-1" yields the term `eclass-1=<a, (eclass-1+(-b))+b>`, which is cyclic, and hence represents arbitrarily (or infinitely) large terms.

More generally, when there is a cyclic term (a term equal to a subterm of itself), rewrite rules which appear not to grow the term can in fact grow the term arbitrarily.

### With cyclic terms, there must be a rule whose LHS has depth at least 2

If all rules have an LHS depth of at most 1, then all rewrites will copy an e-class wholesale. For example: the commutativity rule `a + b => b + a` cannot cause an explosion, as, even if `a = a + b` (so that the term can be written `eclass-1 = <a, eclass-1 + b>`), this could merely rewrite `eclass-1 + b` to `b + eclass-1`and not forcibly expand `eclass-1` to something larger.

### A rule whose LHS overlaps its RHS is trouble

The important thing is actually that there is a composition of rules whose final RHS overlaps the initial LHS, but examples are easier to find for a single rule. Associativity satisfies this condition, as one can apply it twice in a row to a term like `((w+x)+y)+z`. Commutativity also satisfies this condition, but it is also depth-1 (see above).

### Cyclic terms are not necessary

Consider the rule system `A => B`, `f(A, y) => g(B, h(y, y))`, `g(A, x) => f(A, X)`. As a term-rewriting system, it is trivial to see this is terminating. However, as an e-graph rewriting system, the start term `f(A, B)` blows up infinitely. This is because treating it as an e-graph rewriting system effectively adds the rule `B => A`, which renders the term-rewriting system nonterminating (and non-quasi-terminating).

However, the rule set with associativity is still quasi-terminating, but blows up as an e-graph rewriting system. We conjecture that, if R is a quasi-terminating rewrite system which is _symmetric_ in the sense that, for each rule `A => B`,  a corresponding `B => A` rule exists, and R is non-terminating as an e-graph rewriting system, then R can generate a cyclic term.

## Mitigation

No general solution to the problem of infinite derivations from associativity and similar rules is known. There are different domain-specific solutions based on the premise that most applications do not need to find all equivalent terms (saturation), but can make do with merely finding a large number of them (called "moistening" by Edward Kmett).

### Rule scheduling

Egg's default exponential-backoff scheduler ( https://docs.rs/egg/0.6.0/egg/struct.BackoffScheduler.html ) has been found to be effective at mitigating this problem, detecting likely cycles by seeing a rule fire too many times, and then temporarily banning it in favor of other rules.

### Early pruning

For an optimization setting, it can be effective to remove members of an eclass which are strictly worse than another member. For example, if an eclass contains both `0` and `a*0`, then remove the `a*0`.

### Abandon general matching for certain nodes

Chandrakana Nandi writes:

"Speaking of domain specific solutions, in our Szalinski work, we encountered a slightly different flavor of AC-matching due to working with permutations / partitioning of lists. There, we solved the problem by not adding all permutations / partitions in the egraph. Instead we called out to a custom solver to figure out if there is a useful reordering and added just that reordering to the egraph. "

**Comment by mwillsey** [comment]
Thanks for writing this up Jimmy, these are some keen observations! From egg's perspective, we have indeed targeted those applications where saturation isn't required. We were actually just chatting about potentially finding a name for the technique that conveys that we are using e-graphs for rewriting but don't care about saturation ("moistening" wouldn't be my choice 😄, perhaps just "e-graph rewriting").

What kind of applications _require_ saturation? Even when we've used egg as a theorem prover, it's been for undecidable (first-order) logics.

**Reply by remysucre** [reply]
For one, a superoptimizer (e.g. Denali) would require saturation to guarantee optimality.

**Comment by taktoa** [comment]
Isn't the issue with `(a+b)+c => a+(b+c)`, `a+0 => a`, and `(-b)+b => 0` just that equality saturation doesn't actually take an arbitrary TRS as input -- instead, it morally takes a TRS where, for every rule `X => Y`, there exists a rule `Y => X`, and when you give it an arbitrary TRS it is just "completing" the TRS to that form. So really, the condition that matters isn't whether the TRS `T` is terminating, it's whether the _completion_ of `T` is terminating. Am I right in saying that equality saturation always terminates whenever the completion of the input TRS is terminating, or is there some other situation I'm missing?

**Reply by remysucre** [reply]
egg has both bidirectional rules (x <=> y) and directed rules (x => y) which give different results in practice, so I don't think eq. sat. always completes the rules. Also there are rules like a*0=>0 that are impossible to invert.

**Reply by jkoppel** [reply]
>  instead, it morally takes a TRS where, for every rule X => Y, there exists a rule Y => X, and when you give it an arbitrary TRS it is just "completing" the TRS to that form. 

This is not true for several reasons.

First, this doesn't actually work syntactically. TRSs have the restriction that the variables in the RHS must be a subset of those in the LHS. One cannot reverse the rule `a * 0 => 0`, and an e-graph will not associate `0` with every single number. Instead, it's more accurate to say that there is a reversed form of every _specific instantiation_ of a rule that has already fired. For instance, if the instantiation `7 * 0 => 0` runs, then `0 => 7 * 0` is added.

Second, termination depends on the cardinality of eclasses moreso than the cardinality of terms. And thirdly, rule-ordering matters in that you want rules that prove a new term is equivalent to an old eclass to fire before rules that generate new terms.

As a combined example of the second and third, I'll paste here an example I wrote involving De Morgan's law.  Having both `A /\ B => ~(~A \/ ~B)` and `A \/ B => ~(~A /\ ~B)` is clearly nonterminating as a TRS, but can be terminating as an egraph-rewriting system, **so long as the rule `~~A => A` is given priority**.

-----------------------

De Morgan example:


1)

Have:

    ~P /\ ~Q

2) demorgan1 fires.

Have:

    ~P /\ ~Q
     ~(~~P \/ ~~Q)

3) double-negation fires twice:

Have: (using <> for equivalence classes)

    ~P /\ ~Q
    ~(<P, ~~P> \/ <Q, ~~Q>)

4) demorgan2 fires

Have:

    ~P /\ ~Q
    ~(<P, ~~P> \/ <Q, ~~Q>)
    ~<(<P, ~~P> \/ <Q, ~~Q>),    ~(~<P, ~~P> /\ ~<Q, ~~Q>)>

5) double-negation fires

Have:

    ~P /\ ~Q
    ~(<P, ~~P> \/ <Q, ~~Q>)
    ~<(<P, ~~P> \/ <Q, ~~Q>),    ~(<~<P, ~~P>, ~P> /\ <~<Q, ~~Q>, ~Q>)>

The latter reduces to:

    ~<(<P, ~~P> \/ <Q, ~~Q>),    ~(~<P, ~~P> /\ ~<Q, ~~Q>)>

And we are at saturation.

**Reply by taktoa** [reply]
I wasn't claiming that nonterminating TRSes couldn't be terminating as e-graph systems, I was claiming that terminating TRSes whose semantics are that of an equivalence relation always terminate as e-graph systems.

**Reply by mwillsey** [reply]
@jkoppel What do you mean by "is given priority"? In my view, as long as that cancellation rule (or interestingly, its converse)  is present at all it should terminate. Maybe its the distinction between one-at-a-time application and all-at-once application?

**Reply by jkoppel** [reply]
Suppose the De Morgan rules fire twice as often as the cancellation rule. Then there will be intermediate states where the prover believes that `~~~~(A /\ B)`, `~~~~~(A /\ B)`, `~~~~~~(A /\ B)`, etc are all different, and will keep generating new said terms.

(Actually, looking at this now, I'm less sure. But I do know that Yogo marks the double-negation rule as higher priority, which makes a big difference, and I had an example involving unary `f` and `g` symbols that made this more clear.)

What's all-at-once application?

**Reply by mwillsey** [reply]
This may be an egg vs yogo implementation detail. But equality saturation a la egg is defined as
```python
for i in iterations: 
  matches =[]
  for rw in rewrites: # do all the searching
    matches.extend(ematch(rw.lhs))
  for m in matches:   # now mutate the egraph
    apply(m)
```

Fig 5 in the [egg paper](https://dl.acm.org/doi/pdf/10.1145/3434304) shows this as well.

I think I see what you are saying though. If this isn't how you do things, you must be careful to cancel sooner rather than later.

**Reply by remysucre** [reply]
@taktoa I understood you are claiming a sufficient condition and not a necessary condition. But can you make the following precise? 
> TRSes whose semantics are that of an equivalence relation

I didn't understand your initial explanation based on completion for the reason I gave above.

**Reply by jkoppel** [reply]
@mwillsey  I guess that  would solve this instance, but it looks like a superficial defense.  If I split the `~~A => A` rule into two rules, say `~~A => to_eliminate(~~A)` and `to_eliminate(~~A) => A`, then the problem returns.

**Reply by mwillsey** [reply]
@jkoppel I was just about to agree with you, but I tried it out and egg saturates. I even double-checked it's not a saturation-checking bug; you can let it run forever and it's always 16 e-nodes.

```rust
    use egg::*;

    define_language! {
        pub enum Lang {
            "&" = And([Id; 2]),
            "|" = Or([Id; 2]),
            "~" = Not(Id),
            "to-elim" = ToElim(Id),
            Symbol(Symbol),
        }
    }

    let rules = vec![
        rw!("demorgan1";  "(& ?x ?y)"  => "(~ (| (~ ?x) (~ ?y)))"),
        rw!("demorgan2";  "(| ?x ?y)"  => "(~ (& (~ ?x) (~ ?y)))"),
        rw!("cancel1";  "(~ (~ ?x))"  => "(to-elim ?x)"),
        rw!("cancel2";  "(to-elim ?x)"  => "?x"),
    ];

    let runner = Runner::<Lang, (), ()>::default()
        .with_expr(&"(& (~ a) (~ b))".parse().unwrap())
        .with_scheduler(SimpleScheduler)
        .run(&rules);
    assert!(matches!(runner.stop_reason, Some(StopReason::Saturated)));
    assert_eq!(runner.iterations.len(), 5);
    runner.print_report()
```

**Reply by jkoppel** [reply]
Well, I too have had cases where I thought I could write something that my program would fail, but was wrong.

When I gave my example, I intentionally wrote "~~x => to_elim(~~x)" instead of "~~x => to_elim(x)", but I didn't really think it through (e.g.: it changes things that the LHS and RHS of the former rule unify).

I'm going to keep this (whether rule priority matters in terms of infinite-looping even when you apply them simultaneously) flagged in my mind as "unresolved question about equality saturation."

**Reply by mwillsey** [reply]
I think you're right, this is definitely an under-studied problem. One thing we found difficult to articulate in the paper is why all-at-once rule application is "correct" conceptually (as opposed to the traditional search and apply each rule, regardless of order). Perhaps this a useful case study.

FWIW, using the version of to_elim you posted (instead of what I mistakenly used) does not make a difference.

**Reply by jkoppel** [reply]
That's surprising. I would like to understand why that works.

**Reply by chandrakananandi** [reply]
I am not sure how much this has to do with `one-at-a-time application` vs `all-at-once rule application`.

In eqsat, in each round of rule firing, all applicable rules will fire, whether you apply the matches in a batch or right away should not make a difference. Even in the de-morgan example, where is double-negation rule being given more priority than the others?

**Reply by jkoppel** [reply]
Welp, I seem to have been mistaken. Yogo does define some rules to have higher priority, but the double-negation rule is not one of them.

(Yogo's implementation based on the Rete algorithm gives it quite different dynamics from the "rounds" of egg.)

**Reply by chandrakananandi** [reply]
Ah yes I remember now that Yogo uses the Rete based implementation. Thanks for the clarification.

**Reply by remysucre** [reply]
I think @jkoppel's example terminated because of congruence. Even if the egraph contains `~~e`, `~~~~e`, `~~~~~~e`... at some point, one single firing of `~~e => e` will trigger congruence (which runs to a fixpoint) and collapses all of them to a single class.

**Reply by jkoppel** [reply]
That makes sense.

I spent some minutes trying to come up with a new antagonistic example. I think this, or something quite close, will work:

    h(g(x, Z)) =>  h(g(f(x), Z))
    g(f(x), y) => g(x, f(y))
    g(Z, y) => g(Z,Z)

Interpret "Z" as "zero" and "f(f(Z))" as "2", etc.

You start with h(g(i, Z)). Each round, you add h(g(i+1, Z)). It takes i rounds to show that g(i, Z) == g(Z, Z). Congruence does not help. 

Intended outcome: If you let the latter two rules run long enough, then all g(x,y) terms will be merged into the same eclass, and the first rule should stop firing, terminating the system. If you don't, then it runs forever.

**Reply by remysucre** [reply]
Why would the first rule stop firing? It should keep firing until x merges with f(x), which shouldn't happen.

**Reply by jkoppel** [reply]
Start: h(g(Z, Z))

Rule 1 fires, giving h(g(f(Z), Z))

If rules 2 and 3 are allowed to fire to saturation, then g(f(Z), Z) merges with g(Z, Z)

Then, as I understand it: h(g(Z, Z)) and h(g(f(Z), Z)) will now have the same representation h(eclass-1), and rule 1 will not fire because it has already fired on h(eclass-1), right?

If the h(g(x, Z)) LHS will match every distinct x, then we can futz with this some more, say by replacing rule 1 with the rules `h(d(e(g(x, Z)))) =>  h(e(g(f(x), Z)))`  and `h(e(x)) => h(d(e(x)))`, so that the latter rule will most definitely stop firing after merging the g's.

**Reply by remysucre** [reply]
> Then, as I understand it: h(g(Z, Z)) and h(g(f(Z), Z)) will now have the same representation h(eclass-1), and rule 1 will not fire because it has already fired on h(eclass-1), right?

(Sorry I keep editing my answer) Rule 1 didn't fire on h(eclass-1), it fired on h(g(eclass-0, eclass-0)). Therefore it will fire again, this time on h(g(eclass-1, eclass-0)). It will keep firing on h(g(eclass-n, eclass-0)) for increasing n. 

In general rule 2 & 3 can only merge g(Z,Z) with g(n,Z) up to a finite n, whereas rule 1 will keep generating g(n, Z) where n increases without bound. 

I don't yet understand your new rules, but I tend to believe termination is independent of rewrite order: 

Assume saturation terminates for a certain order of rewrite. Then the egraph E after termination represents all possible terms and equivalences given by the rules. Assume for contradiction that saturation doesn't terminate for a different order. Then we will eventually represent all terms and equivalence in E - but that's all the possible terms and equivalences there is, so the egraph must not contain anything else and saturation terminates, contradiction. 

The part "we will eventually represent all terms and equivalence in E" needs a proof but I think it's true because we will eventually fire all the same rules that gave rise to E in the terminating run.

**Reply by remysucre** [reply]
`h(d(e(g(x, Z)))) => h(e(g(f(x), Z)))` and `h(e(x)) => h(d(e(x)))` won't work either because they generate an infinite number of `g(f...f(x), Z)` and we can only merge a finite number of them, for the same reason as above.

**Reply by jkoppel** [reply]
If I understand your argument correctly, it assumes that there is no intermediate state which contains more e-classes than a saturated state, which is false. (There's a similar-looking specious argument that lambda calculus terminates regardless of reduction order: "Eventually all redexes will be reduced.") The crux of my strategy is set up a race between merging eclasses and creating terms, and to make the latter win.

For the second system, here's the trace:

Start: h(e(g(Z, Z)))

Round 1:

h(d(e(g(Z,Z)))

Round 2:

h(e(g(f(Z), Z)))

Running the g rules solo merges g(f(Z), Z) with g(Z, Z).

Now there is only one term of the form `h(e(x))`, and the rule has already fired on it. And now the system should stop. 

Unless I'm really having a major misunderstanding about how this works. (I am, alas, but a second author.)

**Reply by remysucre** [reply]
Indeed, `h(e(x))` won't fire this time, but `h(d(e(g(f(Z), Z))))` is already represented and `h(d(e(g(x, Z)))) => h(e(g(f(x), Z)))` will fire again, and _now_ `h(e(x)) => h(d(e(x)))` fires! 

> it assumes that there is no intermediate state which contains more e-classes than a saturated state

I don't think I need to assume this. There can be more e-classes than after full saturation, but they will be merged in both the terminating run and the hypothetical non-terminating run.

**Reply by jkoppel** [reply]
> Indeed, h(e(x)) won't fire this time, but h(d(e(g(f(Z), Z)))) is already represented and h(d(e(g(x, Z)))) => h(e(g(f(x), Z))) will fire again, and now h(e(x)) => h(d(e(x))) fires!

I see. So, the round-robin rule scheduler that egg uses will cause nontermination. But, in a rule schedule that runs the two `g` rules to saturation before the `h` rules, it will terminate. Therefore, rule ordering affects termination. QED.

(I haven't tried, but this example can be implemented in Yogo. Yogo does not use a round-robin scheduler, but uses the scheduler of the underlying Clara library, which allows one to specify rule "salience" but is otherwise a black box. And, indeed, we've found that specifying salience can make a big difference.)

> There can be more e-classes than after full saturation, but they will be merged in both the terminating run and the hypothetical non-terminating run.

That's fine. My claim is that there can be no upper bound on how many intermediate e-classes will be created, even if the saturated state only has a finite number of e-classes. "All problems are eventually solved" does not imply "There will be a time with no unsolved problems."

**Reply by remysucre** [reply]
> But, in a rule schedule that runs the two g rules to saturation before the h rules, it will terminate.

Except that after the h rules fire again the g rules become unsaturated and should also fire again. Happy to make a phone call, this stuff is tricky.

**Reply by jkoppel** [reply]
Will message you

**Reply by mwillsey** [reply]
Just for my own sake, I ran this in egg and it indeed does not saturate.

**Reply by remysucre** [reply]
I convinced Jimmy that this example should not terminate on the given schedule either. I’ll attempt a formal proof of the independence of termination and order later

**Reply by chandrakananandi** [reply]
Ok I played around a bit with this ruleset:

> R1: h(g(x, Z)) =>  h(g(f(x), Z))
R2: g(f(x), y) => g(x, f(y))
R3: g(Z, y) => g(Z,Z)

and starting with the term `h ( g (Z, Z))`, as soon as R1 applies we get: `h (g (f (Z), Z))` and then R1 again applies to give:
`h (g (f (f (Z)), Z))` and so on.  R2 and R3 rewrite terms down to `g(Z, Z)` but it doesn't stop R1 from continuing to generate the next round of `h (g (f (f (f ...))), Z ))`. So I also agree that this cannot terminate in vanilla equality saturation (I think irrespective of `all-at-once` / `one-at-a-time` application, but I am not 100% on that).

If you stop applying R1 completely at some point and only apply R2 and R3 from there on, it will terminate but that is _not_ what @jkoppel is suggesting below:
> But, in a rule schedule that runs the two g rules to saturation before the h rules, it will terminate.

R1 is allowed to apply again in your schedule, right?

**Reply by chandrakananandi** [reply]
What is rule salience in Clara? Is it a way to stop a rule from being applied if some condition is met?

**Reply by jkoppel** [reply]
> R1 is allowed to apply again in your schedule, right?

Yes it is.

> What is rule salience in Clara? Is it a way to stop a rule from being applied if some condition is met?

http://www.clara-rules.org/docs/conflictsalience/

"Salience is simply a integer property attached to the rule, where rules with higher values will fire before rules with lower values. Here is a simple example:"

It's a fancy word for "priority," because industry people like making up their own terminology.

I am down to <40% confidence that rule priority affects termination, but still at at least 70% confidence it can have a major effect  on running time.

**Comment by ekmett** [comment]
FWIW- Equality "moistening" was originally coined by @deviant-logic. I just adopted it.

**Comment by 0x0f0f0f** [comment]
Hi! Author of Metatheory.jl (re-implementation and extension of egg in julia). Has anything been done on automatic pruning of an e-graph? Im working in parallel towards smarter automatic schedulers. I guess they are two similar ways of reducing the search space, but orthogonally in regard to eclasses for pruning and to rules for scheduling.

**Reply by remysucre** [reply]
I tend to avoid pruning except for perhaps constant folding, since pruning goes against the monotone principle of equality saturation and frequently creates subtle bugs
